### PR TITLE
Added capability to sort IFS files and Members by name or date

### PR DESCRIPTION
**Pull request by sebjulliand**
### Changes
A new submenu has been added to the IFS directories and Physical Source File items in the explorers:
image
image

It allows to sort the content of the selected directory/SPF by name or date.

### Checklist

* [x] have tested my change
* [x] eslint is not complaining

### CODE REVIEW

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@sebjulliand Nice - just gave it a quick spin, and it worked as expected.

Some minor flaws in the design, that I think should be changed / added - or at least discussed:

- In the IFS browser, only directories have the sort menu. I would like it to work on shortcuts as well, since they point to a directory and shows files when opened.
- In the IFS browser, it would be nice to have the sort menu on files as well, so I don't need to scroll up to find the directory and sort.
- In the Object browser, it would be nice to have the sort menu on members as well, like for IFS files.

This is a great step to assist the user in file handling... and you managed to sort members as well...! 😉

**Comment by worksofliam** [comment]
@sebjulliand Before I do this I think it might be worth fixing the conflicts here. Sorry about that!

**Comment by sebjulliand** [comment]
All done @chrjorgensen, you can now sort from members, files and shortcuts.

**Review by chrjorgensen — COMMENTED** [review]
@sebjulliand Code is working just beautifully - only some thoughts about sort on a closed tree view branch: Does it make sense to have the `Sort` menu on a closed branch? It does nothing visually until one opens the branch. Maybe the branch should open if the user selects to sort on the closed branch?

Another thought is about ascending and descending sort. It depends on the situation what the user is interested in: the oldest member / file or the newest member / file. It would be nice to have some kind of selection - or an indicator for the sorting direction, which would switch to the opposite when selected again. E.g. the user selects to sort by date, and the newest are shown first in the list and there is now an `up` arrow in the menu option. Then the user selects it again, and now the oldest are shown first in list and there is now a `down` arrow in the menu option. You know, like in a grid view. Just an idea...

WDYT?

**Comment by sebjulliand** [comment]
@chrjorgensen yes, that makes perfect sense. I'll see what can be done 😁

**Comment by sebjulliand** [comment]
@chrjorgensen, reversing the sort order is easy, but displaying a ⬇ or ⬆ in the command name requires to create separate commands and display one or the other depending on the clicked item's context value which will impact every `when` clause using those items in `package.json`. I'm not quite sure I want to go down this path just to display the current sorting order 😅
Can you think of a better place to display the sorting order? (item description, tooltip...)

**Comment by sebjulliand** [comment]
@chrjorgensen , you're on.
Sorting is now reversed when re-executed and asking to sort on a closed branch will expand the branch.
Let me know what you think 😊

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@sebjulliand Code is still working beautiful, and now with sort direction! 👏 

Only thing discovered is the sort of national characters in name order - and a possible solution is suggested.

I would suggest having the sort order and direction in the item description of the parent (source file/directory), maybe like this:

```
SOURCEFILE (name🔼) description of file
SOURCEFILE (name🔽) description of file
SOURCEFILE (date🔼) description of file
SOURCEFILE (date🔽) description of file
```

File: src/api/IBMiContent.ts [hunk 1]
```diff
@@ -440,27 +439,42 @@ export default class IBMiContent {
...
    * @param remotePath 
    * @return an array of IFSFile
    */
-  async getFileList(remotePath: string): Promise<IFSFile[]> {
+  async getFileList(remotePath: string, order: "name" | "date" = "name", ascending?: boolean): Promise<IFSFile[]> {
     const items: IFSFile[] = [];
 
     const fileListResult = (await this.ibmi.sendCommand({
-      command: `ls -a -p -L ${Tools.escapePath(remotePath)}`
+      command: `ls -a -p -L ${order === "date" ? "-t" : ""} ${(order === 'date' && ascending) || (order === "name" && ascending === false) ? "-r" : ""} ${Tools.escapePath(remotePath)}`
```
[the remark this answers is not in the document]
Comment (sebjulliand) [reply src/api/IBMiContent.ts:477]: You raised an excellent point, thanks. I'll leave it up to `localCompare` to sort by name then.

**Comment by sebjulliand** [comment]
@chrjorgensen there you go:
image

The sorting order will show in the description (only if changed...otherwise it shows on every directory/SPF and the UI gets bloated 😁).

And the file name comparison uses `localCompare` now.
Let me know how it goes!

**Review by chrjorgensen — COMMENTED** [review]
@sebjulliand Perfect, this is so great! I will leave it up to @worksofliam to approve, think he should have a go as well. 😃 

One question: Why change `instance` into a function? Just curious...

**Comment by sebjulliand** [comment]
Thanks @chrjorgensen !
> One question: Why change `instance` into a function? Just curious...

That's because `instance` used to be a constant and it is now a variable (since https://github.com/halcyon-tech/vscode-ibmi/pull/1142). And I found out that if we require the `instance` field outside of a function (at least in JavaScript...I didn't have a problem when importing in TypeScript), we sometimes get the un-initialized variable (depending when the script gets loaded). So to avoid this, I put a function here and there to get the "correct" `instance` field. It won't be necessary once everything will be converted to TS though. 😁

> One small issue: Each time I sort on a directory, a counter and the directory name appears in `DEBUG CONSOLE`

This `console.log` has been here since before I made this change:
image
Maybe we can remove it @worksofliam ?

**Comment by worksofliam** [comment]
@worksofliam Absolutely we can remove it.

**Comment by sebjulliand** [comment]
Removed! @chrjorgensen feel free to approve and merge whenever you want!

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@sebjulliand The code is working fine - except for some new things I noticed:

1. When a source file or directory is closed and I select by name, it sorts in descending order. I would expect it to sort in ascending order first, then descending?
2. I can't create a filter in the Object Browser with this PR! I get the following error when trying: 
billede

**Comment by sebjulliand** [comment]
Thanks @chrjorgensen!
Indeed, recent changes broke the action opening the filter editor.
So that's fixed and now sorting by name on a closes node will sort by name by default.

**Comment by worksofliam** [comment]
@sebjulliand Is this something I can review in place of Christian? Is this really just needed some user testing? That is something I can do :)

**Comment by sebjulliand** [comment]
@worksofliam sure, it just need a final test round 😀

File: src/extension.ts [hunk 2]
```diff
@@ -19,6 +19,8 @@ import { CompileTools } from "./api/CompileTools";
 import { HelpView } from "./views/helpView";
 import { ProfilesView } from "./views/ProfilesView";
 import * as Debug from './api/debug';
+
```
Comment (chrjorgensen) [inline src/extension.ts:22]: Aren't a line like the following missing here?

```suggestion
import IFSBrowser from "./views/ifsBrowser.js";
```

**Review by chrjorgensen — CHANGES_REQUESTED** [review]
@sebjulliand Just gave your last update a spin - but the extension wouldn't even start, but gave this error:
billede

See suggested fix...

**Comment by sebjulliand** [comment]
Hmmm... I pulled and ran npm i, started it and it didn't crash. I'd expect
ObjectBrowser to crash as well then, since it's still JS. I'll dig deeper 😉

**Comment by sebjulliand** [comment]
@chrjorgensen you were right; I guess I wasn't really awake yesterday 😅
Anyway, it's working now!

**Review by worksofliam — APPROVED** [review]
@sebjulliand I tested and I have to say this is really pushing the UX for VS Code - really excellent job! I also secretly really love the use of emoji here! (⬆️ ⬇️)

So, you have my approval. Next steps for this are:

1. Fixing the conflicts and getting it merged.
2. I would really appreciate another PR with tests for your changes to `getFileList` and `getMemberList`

Thanks for everything!

**Comment by sebjulliand** [comment]
Thanks a lot @worksofliam!
Conflicts have been resolved. I'll let you merge and a test PR will follow shortly 😉

**Comment by worksofliam** [comment]
The comparison tests (SQL vs no SQL) for `getMemberList` fails, but to be expected as each types returns a unique `changed` property. I am ok with this, but something to keep in mind when writing new tests @sebjulliand.

image

**Review by worksofliam — APPROVED** [review]
Retested after the merge commit and I am very happy!

Time to release 1.8.0!

**Pull request merged**